File: tier3_advanced/modified_binary_search/solutions/p033_search_rotated.py
"""P033: Search in Rotated Sorted Array (Medium)

TODO: Implement solve() below.
Hint: Determine which half is sorted, then check if target lies in that half.
"""
import sys
sys.path.insert(0, ".")
from src.utils import Problem, TestCase

class Solution(Problem):
    name = "33. Search in Rotated Sorted Array"
    test_cases = [
        TestCase(input=([4,5,6,7,0,1,2], 0), expected=4, label="example 1"),
        TestCase(input=([4,5,6,7,0,1,2], 3), expected=-1, label="example 2"),
        TestCase(input=([1], 0), expected=-1, label="single element"),
    ]

    def solve(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            if nums[left] <= nums[mid]:
                if nums[left] <= target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1

if __name__ == "__main__":
    Solution().run()

File: tier3_advanced/modified_binary_search/solutions/p033_search_rotated.rs
/*
 * P33: Search in Rotated Sorted Array (Medium)
 * https://leetcode.com/problems/search-in-rotated-sorted-array/
 * Topics: Array, Binary Search
 *
 * There is an integer array nums sorted in ascending order (with distinct values).
 * Prior to being passed to your function, nums is possibly rotated at an unknown pivot index k (1 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3 and become [4,5,6,7,0,1,2].
 * Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.
 * You must write an algorithm with O(log n) runtime complexity.
 *
 * Example 1:
 *     Input: nums = [4,5,6,7,0,1,2], target = 0
 *     Output: 4
 *
 * Example 2:
 *     Input: nums = [4,5,6,7,0,1,2], target = 3
 *     Output: -1
 *
 * Example 3:
 *     Input: nums = [1], target = 0
 *     Output: -1
 *
 * Constraints:
 *     - 1 <= nums.length <= 5000
 *     - -10^4 <= nums[i] <= 10^4
 *     - All values of nums are unique.
 *     - nums is an ascending array that is possibly rotated.
 *     - -10^4 <= target <= 10^4
 *
 * Template (python3):
 *     class Solution:
 *         def search(self, nums: List[int], target: int) -> int:
 *
 * Hint: Determine which half is sorted, then check if target lies in that half.
 */

use wasm_libs::*;

impl Solution {
    fn search(nums: &[i32], target: i32) -> i32 {
        let mut left: i32 = 0;
        let mut right: i32 = nums.len() as i32 - 1;
        while left <= right {
            let mid = left + (right - left) / 2;
            if nums[mid as usize] == target {
                return mid;
            }
            if nums[left as usize] <= nums[mid as usize] {
                if nums[left as usize] <= target && target < nums[mid as usize] {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if nums[mid as usize] < target && target <= nums[right as usize] {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        -1
    }
}

struct Solution;

fn main() {
    let nums = read_ints();
    let target = read_int();
    write_int(Solution::search(&nums, target));
    std::process::exit(0);
}
